### PR TITLE
prevent job creation for unresolved plans

### PR DESCRIPTION
**Pull request by MonzElmasry**
- Fix : #96 
- add validation to http response

### CODE REVIEW

File: pkg/upgrade/plan/plan.go [hunk 1]
```diff
@@ -117,7 +117,10 @@ func ResolveChannel(ctx context.Context, url, latestVersion, clusterID string) (
 		}
 		return filepath.Base(redirect.Path), nil
 	}
-	return filepath.Base(url), nil
+	if response.StatusCode < 300 {
```
Comment (galal-hussein) [inline pkg/upgrade/plan/plan.go:120]: I am not sure if we should select any response code less than 300, this is too general

Comment (MonzElmasry) [reply pkg/upgrade/plan/plan.go:120]: i think it's okay if the response is in 200 range except for 302 which is redirected.  
as per our conversation @dweomer

Comment (dweomer) [reply pkg/upgrade/plan/plan.go:120]: Maybe something like `if (response.StatusCode / 200) == 1 {/* return the base of the passed url */}`m what do you think @galal-hussein ?

File: pkg/upgrade/handle_upgrade.go [hunk 2]
```diff
@@ -65,15 +65,17 @@ func (ctl *Controller) handlePlans(ctx context.Context) error {
...
 			}
 			logrus.Debugf("PLAN GENERATING HANDLER: plan=%s/%s@%s, status=%+v", obj.Namespace, obj.Name, obj.ResourceVersion, status)
-			concurrentNodeNames, err := upgradeplan.SelectConcurrentNodeNames(obj, nodes.Cache())
-			if err != nil {
-				return objects, status, err
-			}
-			logrus.Debugf("concurrentNodeNames = %q", concurrentNodeNames)
-			for _, nodeName := range concurrentNodeNames {
-				objects = append(objects, upgradejob.New(obj, nodeName, ctl.Name))
+			if upgradeapiv1.PlanLatestResolved.IsTrue(obj) {
```
Comment (dweomer) [inline pkg/upgrade/handle_upgrade.go:68]: Please instead make the `upgradeapiv1.PlanLatestResolved.IsTrue(obj)` check immediately after the `logrus.Debugf()` line above (which follows the nil check) and reverse it such that if `upgradeapiv1.PlanLatestResolved.IsTrue(obj)` evaluates to `false` then we bail out early and return the same values as the nil check.

File: pkg/upgrade/plan/plan.go (hunk 1 above)
Comment (galal-hussein) [reply pkg/upgrade/plan/plan.go:120]: yeah sounds good, so basically limiting the responses to 200 and 300

File: pkg/upgrade/plan/plan.go [hunk 3]
```diff
@@ -117,7 +117,10 @@ func ResolveChannel(ctx context.Context, url, latestVersion, clusterID string) (
 		}
 		return filepath.Base(redirect.Path), nil
 	}
-	return filepath.Base(url), nil
+	if (response.StatusCode / 200) == 1 {
+		return filepath.Base(url), nil
+	}
+	return filepath.Base(url), fmt.Errorf("invalid url")
```
Comment (dweomer) [inline pkg/upgrade/plan/plan.go:123]:
```suggestion
	return "", fmt.Errorf("unexpected response: %s %s", response.Proto, response.Status)
```

File: e2e/framework/channel.go [hunk 4]
```diff
@@ -7,15 +7,15 @@ import (
 	"os"
 )
 
-func ChannelServer(location string) *httptest.Server {
+func ChannelServer(location string, response int) *httptest.Server {
```
Comment (dweomer) [inline e2e/framework/channel.go:10]: Please refactor `response` to be `status` or `statusCode`.

**Pull request merged**